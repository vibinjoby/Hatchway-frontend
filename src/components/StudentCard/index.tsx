import { FC, useState } from 'react';
import { useDispatch } from 'react-redux';
import styled from 'styled-components';
import Typography from '@material-ui/core/Typography';
import { Collapse } from '@material-ui/core';

import { addTag, StudentState } from '../../features/students/studentsSlice';
import { CssTextField } from '../../pages/home';
import { Button } from '@material-ui/core';

const StudentCard: FC<StudentState> = ({
  city,
  company,
  email,
  firstName,
  grades,
  id,
  lastName,
  pic,
  skill,
  tags,
}: StudentState) => {
  const dispatch = useDispatch();
  const [isSelected, setIsSelected] = useState(false);
  const [tagValue, setTagValue] = useState('');

  const calculateAverage = () => {
    var sum = 0;
    for (var i = 0; i < grades.length; i++) {
      sum += parseFloat(grades[i]);
    }
    return sum / grades.length;
  };

  const DropdownContainer = () => (
    <>
      {grades.map((grade, index) => (
        <GradeContainer key={index}>
          <Typography>Test {index + 1}:</Typography>
          <Typography className="grade">{grade}%</Typography>
        </GradeContainer>
      ))}
    </>
  );

  const Tags = () => {
    return (
      <>
        <TagContainer>
          {tags && tags.length > 0 ? (
            tags.map((tag, index) => (
              <TagWrapper key={index}>
                <Typography>{tag}</Typography>
              </TagWrapper>
            ))
          ) : (
            <></>
          )}
        </TagContainer>
        <CssTextField
          focused
          placeholder="Add a tag"
          className="add-tag-input"
          value={tagValue}
          onChange={e => setTagValue(e.target.value)}
          onKeyPress={e => {
            if (e.keyCode || e.which === 13) {
              dispatch({
                type: addTag.type,
                payload: { data: { tag: tagValue, id } },
              });
              setTagValue('');
            }
          }}
        />
      </>
    );
  };

  return (
    <Container>
      <ImageWrapper src={pic} alt={'Student-pic'} />
      <Wrapper>
        <Typography variant="h4">
          <b>
            {firstName.toUpperCase()} {lastName.toUpperCase()}
          </b>
        </Typography>
        <StudentData>
          <Typography>Email: {email}</Typography>
          <Typography>Company: {company}</Typography>
          <Typography>Skill: {skill}</Typography>
          <Typography>Average: {calculateAverage()}%</Typography>
          {/** Dropdown with grades is shown only when clicked on + */}
          <DropdownWrapper>
            <Collapse in={isSelected}>
              <DropdownContainer />
            </Collapse>
          </DropdownWrapper>
          {/** Show tags horizontally when the user has added tag for each student */}
          {isSelected ? Tags() : <></>}
        </StudentData>
      </Wrapper>
      <Button
        disableRipple
        className="expand-btn"
        onClick={() => setIsSelected(val => !val)}>
        {!isSelected ? '+' : '-'}
      </Button>
    </Container>
  );
};

const Container = styled.section`
  display: flex;
  width: 100%;
  flex-direction: row;
  align-items: flex-start;
  .expand-btn {
    position: absolute;
    cursor: pointer;
    font-size: 50px;
    font-weight: bold;
    color: grey;
    height: 0em;
    margin-left: auto;
    margin-top: 10px;
    right: 0;
    background: transparent;
  }
  .expand-btn:hover {
    background-color: #fff;
  }
`;

const ImageWrapper = styled.img`
  width: 80px;
  height: 80px;
  border: 1px solid grey;
  border-radius: 40px;
  margin: 10px 0;
`;

const Wrapper = styled.div`
  margin: 0 40px;
`;

const StudentData = styled.div`
  display: flex;
  flex-direction: column;
  margin: 0 12px;
  span {
    margin: 2px 0;
  }

  .add-tag-input {
    margin: 10px 0;
  }
`;

const DropdownWrapper = styled.div`
  margin: 20px 0 0 0;
`;

const GradeContainer = styled.div`
  display: flex;
  flex-direction: row;
  .grade {
    margin-left: 22px;
  }
`;

const TagContainer = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
`;

const TagWrapper = styled.div`
  background-color: #dcdcdc;
  margin-right: 10px;
  margin-top: 10px;
  padding: 5px;
  border-radius: 5px;
`;

export default StudentCard;
