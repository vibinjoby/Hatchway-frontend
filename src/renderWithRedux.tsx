import { render } from '@testing-library/react';
import { Provider } from 'react-redux';

import { store } from './store/store';

const renderWithRedux = (component: any) => {
  return {
    ...render(<Provider store={store}>{component}</Provider>),
  };
};

export default renderWithRedux;
