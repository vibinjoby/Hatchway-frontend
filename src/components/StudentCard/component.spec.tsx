import '@testing-library/jest-dom/extend-expect';
import { fireEvent, screen } from '@testing-library/react';

import StudentCard from './index';
import renderWithRedux from '../../renderWithRedux';

test('renders a message', () => {
  const MOCK_IMAGE =
    'https://storage.googleapis.com/hatchways-app.appspot.com/assessments/data/frontend/images/voluptasdictablanditiis.jpg';
  const MOCK_EMAIL = 'iorton0@imdb.com';
  const MOCK_COMPANY = 'Yadel';
  const MOCK_SKILL = 'Oracle';
  const MOCK_FIRSTNAME = 'Ingaberg';
  const MOCK_LASTNAME = 'Orton';
  const MOCK_GRADES = ['78', '100', '92', '86', '89', '88', '91', '87'];

  const { container, getByText } = renderWithRedux(
    <StudentCard
      city={'FushÃ«-Muhurr'}
      company={MOCK_COMPANY}
      email={MOCK_EMAIL}
      firstName={MOCK_FIRSTNAME}
      grades={MOCK_GRADES}
      id={'1'}
      lastName={MOCK_LASTNAME}
      pic={MOCK_IMAGE}
      skill={MOCK_SKILL}
      tags={[]}
    />,
  );
  //Match the snapshot for content change
  expect(container.firstChild).toMatchSnapshot();
  //Check if all student data is rendered
  expect(
    getByText(`${MOCK_FIRSTNAME.toUpperCase()} ${MOCK_LASTNAME.toUpperCase()}`),
  ).toBeTruthy();

  expect(getByText(`Email: ${MOCK_EMAIL}`)).toBeTruthy();
  expect(getByText(`Company: ${MOCK_COMPANY}`)).toBeTruthy();
  expect(getByText(`Skill: ${MOCK_SKILL}`)).toBeTruthy();

  //Should be able to expand contents with button press
  fireEvent.click(screen.getByText('+'));

  //Check if all grades are visible when expanded
  MOCK_GRADES.map((grade, index) => {
    expect(getByText(`Test ${index + 1}:`)).toBeTruthy();
    expect(getByText(`${grade}%`)).toBeTruthy();
  });

  //Should be able to hide the grades
  fireEvent.click(screen.getByText('-'));
});
