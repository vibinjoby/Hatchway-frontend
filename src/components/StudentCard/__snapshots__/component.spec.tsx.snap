// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a message 1`] = `
<section
  class="sc-dlnjwi hPKiOZ"
>
  <img
    alt="Student-pic"
    class="sc-hKFxyN faIHOp"
    src="https://storage.googleapis.com/hatchways-app.appspot.com/assessments/data/frontend/images/voluptasdictablanditiis.jpg"
  />
  <div
    class="sc-eCApnc klaond"
  >
    <h4
      class="MuiTypography-root MuiTypography-h4"
    >
      <b>
        INGABERG
         
        ORTON
      </b>
    </h4>
    <div
      class="sc-jSFjdj fhvhJV"
    >
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Email: 
        iorton0@imdb.com
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Company: 
        Yadel
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Skill: 
        Oracle
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1"
      >
        Average: 
        88.875
        %
      </p>
      <div
        class="sc-gKAaRy iYHbGS"
      >
        <div
          class="MuiCollapse-root MuiCollapse-hidden"
          style="min-height: 0px;"
        >
          <div
            class="MuiCollapse-wrapper"
          >
            <div
              class="MuiCollapse-wrapperInner"
            >
              <div
                class="sc-iCoGMd bGxwXD"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  Test 
                  1
                  :
                </p>
                <p
                  class="MuiTypography-root grade MuiTypography-body1"
                >
                  78
                  %
                </p>
              </div>
              <div
                class="sc-iCoGMd bGxwXD"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  Test 
                  2
                  :
                </p>
                <p
                  class="MuiTypography-root grade MuiTypography-body1"
                >
                  100
                  %
                </p>
              </div>
              <div
                class="sc-iCoGMd bGxwXD"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  Test 
                  3
                  :
                </p>
                <p
                  class="MuiTypography-root grade MuiTypography-body1"
                >
                  92
                  %
                </p>
              </div>
              <div
                class="sc-iCoGMd bGxwXD"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  Test 
                  4
                  :
                </p>
                <p
                  class="MuiTypography-root grade MuiTypography-body1"
                >
                  86
                  %
                </p>
              </div>
              <div
                class="sc-iCoGMd bGxwXD"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  Test 
                  5
                  :
                </p>
                <p
                  class="MuiTypography-root grade MuiTypography-body1"
                >
                  89
                  %
                </p>
              </div>
              <div
                class="sc-iCoGMd bGxwXD"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  Test 
                  6
                  :
                </p>
                <p
                  class="MuiTypography-root grade MuiTypography-body1"
                >
                  88
                  %
                </p>
              </div>
              <div
                class="sc-iCoGMd bGxwXD"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  Test 
                  7
                  :
                </p>
                <p
                  class="MuiTypography-root grade MuiTypography-body1"
                >
                  91
                  %
                </p>
              </div>
              <div
                class="sc-iCoGMd bGxwXD"
              >
                <p
                  class="MuiTypography-root MuiTypography-body1"
                >
                  Test 
                  8
                  :
                </p>
                <p
                  class="MuiTypography-root grade MuiTypography-body1"
                >
                  87
                  %
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-text expand-btn"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      +
    </span>
  </button>
</section>
`;
